### ISSUE TITLE
Having a problem designing/calculating LP capacity of an altar

### ISSUE THREAD

**Issue by the-creamster** [WayofTime/BloodMagic#1781]
#### Issue Description:
This is more of a question than an issue as i'm returning to modded after some time of not playing it. I know how the mod works as i used this one heavily but it's changed sufficiently that i'm left trying to figure out what is now optimal.

#### What happens:
The information given in the book states percentages to do calculations with but some of mine aren't matching what is displayed in-game(was getting values much greater than 271k compared to in-game). Also have i overused some runes/under used others??

Pics of my test build so far:
![2021-02-08_14 05 23](https://user-images.githubusercontent.com/2037833/107176308-b5feca00-6a1a-11eb-8d86-4d68ffca9896.png)
![2021-02-08_14 05 31](https://user-images.githubusercontent.com/2037833/107176323-be570500-6a1a-11eb-9ee2-aa19f743077e.png)

____
#### Affected Versions (Do *not* use "latest"):
FTB custom client Direwolf20 1.16 modpack v1.6.1

**Comment by wrincewind** [comment]
When you say 'values greater than 271k' what exactly are you referring to? How many of which runes are you using, what are you expecting and what are you getting?

**Comment by the-creamster** [comment]
Well based on the information given in the book and other previous posts from older versions(even if the exact numbers are now different) some people have been able to hit 400k+ and my calculations so far which i get the feeling are a tad wrong have reached 500k+ LP max in the altar. My question is more specifically is the current pic of the test build altar good enough(this will be the crafting/orb filling one) or are there glaring inefficiencies that are hurting me like too many capacity runes and not enough augmented capacity runes etc...

I can roll with that one as i'm happy with it but if i can squeeze another 50-150k more LP in storage then i'll be happier. As i said i've not played modded for around 18 months so a tad of of touch with the intricacies of the systems at play that's all.

**Comment by BlueAgent** [comment]
Currently the information about [augmented capacity runes in the book](https://github.com/WayofTime/BloodMagic/blob/5d33f99d47a4630acc62cb1fe762674dc1f3b9e4/src/main/resources/data/bloodmagic/patchouli_books/guide/en_us/entries/altar/blood_rune/aug_capacity_rune.json#L8) is out of date.

The current (as of this post) formula can be found in [PR#1719](https://github.com/WayofTime/BloodMagic/pull/1719) or in the [code](https://github.com/WayofTime/BloodMagic/blob/5d33f99d47a4630acc62cb1fe762674dc1f3b9e4/src/main/java/wayoftime/bloodmagic/altar/BloodAltar.java#L460-L462).

For a spreadsheet that has the optimal combinations of runes for a given amount of runes you want to dedicated to capacity see [here](https://docs.google.com/spreadsheets/d/1amkQyyG35ugstlj5k365wNtiwzxSWvcfmvLywyRQkD8/edit#gid=0) and look at the column labeled `Merged (PR#1719)`. The `Initial Port (1.16)` column should match your calculations going of the book's information, but might have to multiply by the base volume of the altar (10 buckets).

Compared to the older versions (1.12.2 and before) you have to use a mix (no easy way to describe, but it approaches a 50/50 split at the limit) of runes and the efficiency of capacity runes in general were buffed for lower amounts of runes dedicated to capacity (`7<=runes<=29`) and nerfed for higher amounts (`30<=runes`) dedicated to capacity.

**Comment by WayofTime** [comment]
I am going to take a closer look at it. Honestly, having a formula that requires a simulation tool feels a bit too much, so I'll try to find a good solution in the coming days.

**Comment by the-creamster** [comment]
Thanks so much both of you as this clears the doubt i was having. Needless to say i won't be touching BM on server for a little bit; i still got plenty of other mods to focus on to keep me busy. I'll keep it open incase others want to chime in/you want to re-iterate facts etc...

Edit:
While your looking into it do the other rune calculations actually work as intended??:
- Rune of the Orb
- Charging rune
- self/sacrifice
- displacement(how much is the in/out tranfer rate affected by this one)
- speed

**Comment by WayofTime** [comment]
The only thing that was changed was the Capacity formula, and that was pulled too early. Everything else functions as stated in the book.

**Comment by BlueAgent** [comment]
TL;DR;
Do note this is just a suggestion. The goal of having it be easier to calculate is good. Even just using the old formula would be okay but that wouldn't have a mix of runes so this keeps it the same but requires a mix.

- Scaling exactly the same as old (1.12.2 and before) 
- Matches intent to use capacity runes instead of switching to all augmented
- 0 to 14 runes all capacity
- 15 and onwards, even split of capacity and augmented capacity runes (+1 augmented rune for odd numbers)
- Can be described (see below for description)

Augmented capacity runes do look really nice though.

Posting here for posterity (originally on discord.):

---

Got a suggestion for the capacity formula for the altar.
It keeps exactly the same optimal values as the older 1.12.2 version of blood magic for a given number of runes, but requires the use of a mix of capacity and augmented runes as intended.
After 14 runes, optimal becomes an even split of cap and augmented runes (for odd amounts, +1 aug rune)
```
orig              = 0.2 * cap + 1.1 ^ aug
init116           = (1 + 0.2 * cap) * 1.075 ^ aug
current (PR#1719) = (1 + 0.2 * cap) * 1.1 ^ (aug * 0.99 ^ abs(aug - cap))
proposed          = 0.2 * max(0, cap - aug) + 1.1 ^ min(2 * aug, 2 * cap + 1)
proposed (equiv.) = 0.2 * max(0, cap - aug) + 1.21 ^ min(aug, cap + 0.5)
```
For the book description it would be challenging but something like:
The Augmented Capacity Rune increases the capacity of the altar by 121%
However, each Augmented Capacity Rune will disable one capacity rune in order to empower itself.
You can have one Augmented Capacity Rune powered by the altar itself but it will only work at half effectiveness (110%).

Compare here (check out graph and data sheets at bottom): https://docs.google.com/spreadsheets/d/1amkQyyG35ugstlj5k365wNtiwzxSWvcfmvLywyRQkD8/edit#gid=0

**Comment by wrincewind** [comment]
I dunno - looking over it again, that's just replaced 'use 7 of these and then as many of those as you want' with 'use equal numbers of these and those'. It's a lot more complicated, but is it any more engaging? is it more fun?

**Comment by BlueAgent** [comment]
To clarify:
- 1.12.2 and earlier: Normal for 1 to 14, then switch to all augmented onwards
- Intial 1.16 port: Normal for 1 to 9, then use 9 normal and the rest augmented
- **Currently implemented** (PR #1719): Normal for 1 to 6, then after it gets complicated. It's not just use 7 normal then all augmented (see the PR or the spreadsheet).
- Proposed: Normal for 1 to 15, then after that use a 50/50 split with odd numbers being an extra augmented rune.

Don't think it's that much more engaging to have to use a mix of runes, but IIRC that was what was intended with the intial port / changes. I agree with you though. Thought it was fine before in 1.12.2 and earlier (the augmented runes honestly look super cool too), more interested in the per rune slot efficiency changes.

Just hoping that the high end usage of capacity runes isn't reduced too much and the lower amounts of capacity runes don't get buffed. Since you don't really need that much capacity unless you're doing self-sacrifice, which is much more interesting than mob sacrifice IMO as there's a bunch of things you can do to boost it up (ofc requiring the capacity for it). Increasing the power of lower number of runes and reducing the power of higher number of runes (as it is currently) is overally pushing towards mob sacrifice rather than self-sacrifice. Mob sacrifice is automatable even while offline so its already got that as a huge advantage. This is however a bit off topic though.
